commons/src/main/java/org/infinispan/commons/configuration/attributes/AttributeSet.java

```diff
@@ -0,0 +1,196 @@
+package org.infinispan.commons.configuration.attributes;
+
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.infinispan.commons.logging.Log;
+import org.infinispan.commons.logging.LogFactory;
+
+/**
+ * AttributeSet is a container for {@link Attribute}s. It is constructed by passing in a list of {@link AttributeDefinition}s.
+ * AttributeSets are initially unprotected, which means that the contained attributes can be modified. If the {@link #protect()} method is invoked
+ * then only attributes which are not {@link AttributeDefinition#isImmutable()} can be modified from then on.
+ *
+ * @author Tristan Tarrant
+ * @since 7.2
+ */
+public class AttributeSet implements AttributeListener<Object> {
+   private static final Log log = LogFactory.getLog(AttributeSet.class);
+   private final String name;
+   private final Map<String, Attribute<? extends Object>> attributes;
+   private boolean protect;
+
+   @SafeVarargs
+   public AttributeSet(Class<?> klass, AttributeDefinition<?>... attributeDefinitions) {
+      this(klass.getSimpleName(), attributeDefinitions);
+   }
+
+   @SafeVarargs
+   public AttributeSet(String name, AttributeDefinition<?>... attributeDefinitions) {
+      this.name = name;
+      this.attributes = new LinkedHashMap<>(attributeDefinitions.length);
+      for (AttributeDefinition<?> def : attributeDefinitions) {
+         if (attributes.containsKey(def.name())) {
+            throw log.attributeSetDuplicateAttribute(def.name(), name);
+         }
+         Attribute<Object> attribute = (Attribute<Object>) def.toAttribute();
+         if (!attribute.isImmutable())
+            attribute.addListener(this);
+         this.attributes.put(def.name(), attribute);
+      }
+   }
+
+   @SafeVarargs
+   public AttributeSet(Class<?> klass, AttributeSet attributeSet, AttributeDefinition<?>... attributeDefinitions) {
+      this(klass.getSimpleName(), attributeSet, attributeDefinitions);
+   }
+
+   public AttributeSet(String name, AttributeSet attributeSet, AttributeDefinition<?>[] attributeDefinitions) {
+      this.name = name;
+      this.attributes = new LinkedHashMap<>(attributeDefinitions.length + attributeSet.attributes.size());
+      for (Attribute<? extends Object> attribute : attributeSet.attributes.values()) {
```

Comment:
Same thing here, `?` is enough.

Time to use a modern IDE which shows you this right in your screen? :tongue: :wink: